src/entity/Living.php

```diff
@@ -721,19 +722,32 @@ protected function applyFrostWalker(int $level) : void{
 		$y = $this->location->getFloorY() - 1;
 		$baseZ = $this->location->getFloorZ();
 
-		$frostedIce = VanillaBlocks::FROSTED_ICE();
+		$cancelled = false;
+		$liquid = VanillaBlocks::WATER();
+		$targetBlock = VanillaBlocks::FROSTED_ICE();
+		if(EntityFrostWalkerEvent::hasHandlers()){
+			$ev = new EntityFrostWalkerEvent($this, $radius, $liquid, $targetBlock);
+			$ev->call();
+			$cancelled = $ev->isCancelled();
+			$radius = $ev->getRadius();
+			$liquid = $ev->getLiquid();
+			$targetBlock = $ev->getTargetBlock();
+		}
+		if($cancelled){
+			return;
+		}
 		for($x = $baseX - $radius; $x <= $baseX + $radius; $x++){
 			for($z = $baseZ - $radius; $z <= $baseZ + $radius; $z++){
 				$block = $world->getBlockAt($x, $y, $z);
 				if(
-					!$block instanceof Water ||
+					!$block instanceof $liquid ||
```

Comment:
This shouldn't be using instanceof. Prefer isSameState() instead and that way you also don't need to do an isSource check.